core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,265 @@
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.env.ShardLock;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.gateway.MetaStateService;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.index.shard.ShardPath;
+import org.elasticsearch.index.shard.ShardStateMetaData;
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+/**
+ * Renames index folders from {index.name} to {name}_{index.uuid}
+ */
+public class IndexFolderUpgrader {
+    private final MetaDataStateFormat<IndexMetaData> indexStateFormat;
+    private final NodeEnvironment nodeEnv;
+    private final static ESLogger logger = Loggers.getLogger(IndexFolderUpgrader.class);
+
+
+    /**
+     * Creates a new upgrader instance
+     * @param nodeEnv the node env to operate on
+     * @param indexStateFormat format to read state
+     */
+    IndexFolderUpgrader(NodeEnvironment nodeEnv, MetaDataStateFormat<IndexMetaData> indexStateFormat) {
+        this.indexStateFormat = indexStateFormat;
+        this.nodeEnv = nodeEnv;
+    }
+
+    /**
+     * Copies shard content from <code>shardPaths</code> to <code>targetShardPath</code>. All content
+     * from <code>shardPaths</code> are deleted after being copied
+     */
+    void upgradeShard(final ShardPath targetShardPath, final Path[] shardPaths) throws IOException {
+        logger.info("{} upgrading index from {}", targetShardPath.getShardId(), shardPaths);
+        final ShardStateMetaData loaded = ShardStateMetaData.FORMAT.loadLatestState(logger, shardPaths);
+        if (loaded == null) {
+            throw new IllegalStateException(targetShardPath.getShardId() + " no shard state found in any of: " +
+                Arrays.toString(shardPaths));
+        }
+        logger.info("{} loaded shard state {}",  targetShardPath.getShardId(), loaded);
+        ShardStateMetaData.FORMAT.write(loaded, loaded.legacyVersion, targetShardPath.getShardStatePath());
+        final Path indexPath = targetShardPath.resolveIndex();
+        Files.createDirectories(indexPath);
+        try (SimpleFSDirectory directory = new SimpleFSDirectory(indexPath)) {
+            try (final Lock lock = directory.obtainLock(IndexWriter.WRITE_LOCK_NAME)) {
+                upgradeFiles(targetShardPath, indexPath, ShardPath.INDEX_FOLDER_NAME, shardPaths);
```

Comment:
why do we need all these file copies? can't we just do a top level folder rename? consolidating to one folder is a 2.0 feature?